### PR TITLE
[Feature] Support VoxelPose

### PR DESCRIPTION
**Pull request by wusize**
## Motivation

Support VoxelPose Algorithm in MMPose.

## Modification

1. modify `.dev_scripts/github/update_model_index.py` to support multi-view 3d pose.
2. Add `mmpose/core/camera/single_camera_torch.py` to apply camera-transformation in pytorch.
3. Add `affine_transform_torch` in `mmpose/core/post_processing/post_transforms.py`.
4. Add `MultiItemProcess` and `MultiItemDeduplicate` in `mmpose/datasets/pipelines/shared_transform.py`
5. Delete unused variables (img) in `mmpose/datasets/pipelines/bottom_up_transform.py`.
6. Add config, dataset, models/detector, models/head, models/backbone and related docs.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpose/datasets/pipelines/shared_transform.py [hunk 1]
```diff
@@ -406,6 +406,44 @@ def __repr__(self):
         return repr_str
 
 
+@PIPELINES.register_module()
+class MultiItemProcess:
+    """
```
Comment (jin-s13) [inline mmpose/datasets/pipelines/shared_transform.py:411]: Please complete this description

File: mmpose/datasets/pipelines/pose3d_transform.py [hunk 2]
```diff
@@ -540,3 +540,91 @@ def __call__(self, results):
         results['target'] = target
         results['target_weight'] = target_weight
         return results
+
+
+@PIPELINES.register_module()
+class GenerateVoxel3DHeatmapTarget:
```
Comment (jin-s13) [inline mmpose/datasets/pipelines/pose3d_transform.py:546]: Unittests are needed.

**Comment by jin-s13** [comment]
Need doc/log/ckpt.

File: mmpose/datasets/pipelines/pose3d_transform.py (hunk 2 above)
Comment (wusize) [reply mmpose/datasets/pipelines/pose3d_transform.py:546]: Ok

File: mmpose/models/heads/voxelpose_head.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+        topk_values, topk_unravel_index = nms(root_cubes.detach(),
+                                              self.num_cand)
+        topk_unravel_index = self.get_real_loc(topk_unravel_index)
+
+        grid_centers = torch.zeros(
+            batch_size, self.num_cand, 5, device=root_cubes.device)
+        grid_centers[:, :, 0:3] = topk_unravel_index
+        grid_centers[:, :, 4] = topk_values
+        # grid_centers[:, :, 3] = (topk_values > self.threshold).float() - 1.0
```
Comment (jin-s13) [inline mmpose/models/heads/voxelpose_head.py:70]: del it if unncessary.

**Comment by wusize** [comment]
@ly015 Please help review the codes, especially the modification #1 to #5.

File: configs/body/3d_kpt_mview_rgb_img/README.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# Multi-view 3D Human Body Pose Estimation
+
+3D pose estimation is the prediction of X, Y, Z coordinates of human body joints from multi-view RGB images.
```
Comment (jin-s13) [inline configs/body/3d_kpt_mview_rgb_img/README.md:3]: Multi-view 3D human body pose estimation targets at predicting the X, Y, Z coordinates of human body joints from multi-view RGB images.

File: configs/body/3d_kpt_mview_rgb_img/voxelpose/README.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  title={VoxelPose: Towards Multi-Camera 3D Human Pose Estimation in Wild Environment},
+  author={Tu, Hanyue and Wang, Chunyu and Zeng, Wenjun},
+  booktitle={ECCV},
+  year={2020}
+}
+```
+
+</details>
+
+VoxelPose proposes to break down the task of 3d human pose estimation into 2 stages: (1) Human center detection by Cuboid Proposal Network
```
Comment (jin-s13) [inline configs/body/3d_kpt_mview_rgb_img/voxelpose/README.md:19]: Human center detection (Cuboid Proposal Network)

File: mmpose/datasets/datasets/base/kpt_3d_mview_rgb_img_direct_dataset.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+import json_tricks as json
+import numpy as np
+from torch.utils.data import Dataset
+
+from mmpose.datasets import DatasetInfo
+from mmpose.datasets.pipelines import Compose
+
+
+class Kpt3dMviewRgbImgDirectDataset(Dataset, metaclass=ABCMeta):
+    """Base class for keypoint 3D top-down pose estimation with single-view RGB
```
Comment (jin-s13) [inline mmpose/datasets/datasets/base/kpt_3d_mview_rgb_img_direct_dataset.py:14]: Multiview

File: mmpose/datasets/datasets/body3d/body3d_mview_direct_panoptic_dataset.py [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+from mmcv import Config
+
+from mmpose.core.camera import SimpleCamera
+from mmpose.datasets.builder import DATASETS
+from mmpose.datasets.datasets.base import Kpt3dMviewRgbImgDirectDataset
+
+
+@DATASETS.register_module()
+class Body3DMviewDirectPanopticDataset(Kpt3dMviewRgbImgDirectDataset):
+    """Panoptic dataset for top-down hand pose estimation.
```
Comment (jin-s13) [inline mmpose/datasets/datasets/body3d/body3d_mview_direct_panoptic_dataset.py:22]: body

File: mmpose/datasets/datasets/body3d/body3d_mview_direct_panoptic_dataset.py [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+
+    `Hand Keypoint Detection in Single Images using Multiview
```
Comment (jin-s13) [inline mmpose/datasets/datasets/body3d/body3d_mview_direct_panoptic_dataset.py:24]: this is not correct

File: mmpose/datasets/datasets/body3d/body3d_mview_direct_panoptic_dataset.py [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+        """
+        self.num_joints = data_cfg['num_joints']
+        assert self.num_joints <= 19
+        self.seq_list = data_cfg['seq_list']
+        self.cam_list = data_cfg['cam_list']
+        self.num_cameras = data_cfg['num_cameras']
+        assert self.num_cameras == len(self.cam_list)
+        self.seq_frame_interval = data_cfg.get('seq_frame_interval', 1)
+        self.subset = data_cfg.get('subset', 'train')
+        # self.need_2d_label = data_cfg.get('need_2d_label', False)
```
Comment (jin-s13) [inline mmpose/datasets/datasets/body3d/body3d_mview_direct_panoptic_dataset.py:135]: del it

File: mmpose/datasets/datasets/body3d/body3d_mview_direct_panoptic_dataset.py [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+                'sequence_list': self.seq_list,
+                'interval': self.seq_frame_interval,
+                'cam_list': self.cam_list,
+                'db': self.db
+            }
+            with open(self.db_file, 'wb') as f:
+                pickle.dump(info, f)
+
+        self.db_size = len(self.db)
+        # self.db = self._get_db()
```
Comment (jin-s13) [inline mmpose/datasets/datasets/body3d/body3d_mview_direct_panoptic_dataset.py:118]: del it

File: mmpose/models/heads/voxelpose_head.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+# ------------------------------------------------------------------------------
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from ..builder import HEADS
+
+
+def get_index(indices, shape):
```
Comment (jin-s13) [inline mmpose/models/heads/voxelpose_head.py:14]: need docstring

File: mmpose/models/detectors/voxelpose.py [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+        if isinstance(self.heatmap_size, int):
+            self.heatmap_size = [self.heatmap_size, self.heatmap_size]
+
+    def compute_grid(self, box_size, box_center, num_bins, device=None):
+        if isinstance(box_size, int) or isinstance(box_size, float):
+            box_size = [box_size, box_size, box_size]
+        if isinstance(num_bins, int):
+            num_bins = [num_bins, num_bins, num_bins]
+
+        grid1Dx = torch.linspace(
```
Comment (jin-s13) [inline mmpose/models/detectors/voxelpose.py:75]: Does `grid_1Dx` look better?

File: mmpose/models/detectors/voxelpose.py [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+
+    def forward(self, cubes):
+        return self.v2v_net(cubes)
+
+
+class ProjectLayer(nn.Module):
+
+    def __init__(self, cfg):
+        """Project layer to get voxel feature. Adapted from
+        https://github.com/microsoft/voxelpose-
```
Comment (jin-s13) [inline mmpose/models/detectors/voxelpose.py:53]: ""

File: mmpose/models/backbones/v2v_net.py [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+            in_channels,
+            out_channels,
+            kernel_size,
+            stride=1,
+            padding=((kernel_size - 1) // 2),
+            conv_cfg=conv_cfg,
+            norm_cfg=norm_cfg,
+            bias=True)
+
+    def forward(self, x):
```
Comment (jin-s13) [inline mmpose/models/backbones/v2v_net.py:45]: add docstring.

File: mmpose/datasets/datasets/body3d/body3d_mview_direct_panoptic_dataset.py [hunk 15]
```diff
@@ -0,0 +1,473 @@
...
+
+        precise = np.concatenate(([0], precise, [0]))
+        recall = np.concatenate(([0], recall, [1]))
+        index = np.where(recall[1:] != recall[:-1])[0]
+        ap = np.sum((recall[index + 1] - recall[index]) * precise[index + 1])
+
+        return ap, recall[-2]
+
+    @staticmethod
+    def _eval_list_to_mpjpe(eval_list, threshold=500):
```
Comment (jin-s13) [inline mmpose/datasets/datasets/body3d/body3d_mview_direct_panoptic_dataset.py:433]: add docstring

**Comment by jin-s13** [comment]
Need to rebase to dev-0.21  @wusize
